Framework/MDAlgorithms/src/LoadDNSSCD.cpp

```diff
@@ -0,0 +1,606 @@
+            static_cast<coord_t>(extentMaxs[i]), 5)));
+  }
+
+  // Set coordinate system
+  m_OutWS->setCoordinateSystem(coordinateSystem);
+
+  // calculate RUB matrix
+  Mantid::Geometry::OrientedLattice o;
+  o = Mantid::Geometry::OrientedLattice(a, b, c, alpha, beta, gamma);
+  o.setUFromVectors(Mantid::Kernel::V3D(u[0], u[1], u[2]),
+                    Mantid::Kernel::V3D(v[0], v[1], v[2]));
+
+  double omega_offset = getProperty("OmegaOffset");
+  omega_offset *= -1.0 * deg2rad;
+  DblMatrix rotm(3, 3);
+  rotm[0][0] = std::cos(omega_offset);
+  rotm[0][1] = 0.0;
+  rotm[0][2] = std::sin(omega_offset);
+  rotm[1][0] = 0.0;
+  rotm[1][1] = 1.0;
+  rotm[1][2] = 0.0;
+  rotm[2][0] = -std::sin(omega_offset);
+  rotm[2][1] = 0.0;
+  rotm[2][2] = std::cos(omega_offset);
+
+  DblMatrix ub(o.getUB());
+  ub = rotm * ub;
+  o.setUB(ub);
+  DblMatrix ub_inv(ub);
+  // invert the UB matrix
+  ub_inv.Invert();
+
+  // Creates a new instance of the MDEventInserter to output workspace
+  MDEventWorkspace<MDEvent<3>, 3>::sptr mdws_mdevt_3 =
+      boost::dynamic_pointer_cast<MDEventWorkspace<MDEvent<3>, 3>>(m_OutWS);
+  MDEventInserter<MDEventWorkspace<MDEvent<3>, 3>::sptr> inserter(mdws_mdevt_3);
+
+  // create a normalization workspace
+  IMDEventWorkspace_sptr normWS = m_OutWS->clone();
+
+  // Creates a new instance of the MDEventInserter to norm workspace
+  MDEventWorkspace<MDEvent<3>, 3>::sptr normws_mdevt_3 =
+      boost::dynamic_pointer_cast<MDEventWorkspace<MDEvent<3>, 3>>(normWS);
+  MDEventInserter<MDEventWorkspace<MDEvent<3>, 3>::sptr> norm_inserter(
+      normws_mdevt_3);
+
+  // scattering angle limits
+  std::vector<double> tth_limits = getProperty("TwoThetaLimits");
+  double theta_min = tth_limits[0] * deg2rad / 2.0;
+  double theta_max = tth_limits[1] * deg2rad / 2.0;
+
+  // Go though each element of m_data to convert to MDEvent
+  for (ExpData ds : m_data) {
+    uint16_t runnumber = 1;
+    signal_t norm_signal(ds.norm);
+    signal_t norm_error = std::sqrt(m_normfactor * norm_signal);
+    double k = 2.0 / ds.wavelength;
+    for (size_t i = 0; i < ds.detID.size(); i++) {
+      signal_t signal(ds.signal[i]);
+      signal_t error = std::sqrt(signal);
+      detid_t detid(ds.detID[i]);
+      double theta = 0.5 * (ds.detID[i] * 5.0 - ds.deterota) * deg2rad;
+      if ((theta > theta_min) && (theta < theta_max)) {
+        double omega = (ds.huber - ds.deterota) * deg2rad - theta;
+        V3D uphi(-cos(omega), 0, -sin(omega));
+        V3D hphi = uphi * k * sin(theta);
+        V3D hkl = ub_inv * hphi;
+        std::vector<Mantid::coord_t> millerindex(3);
+        millerindex[0] = static_cast<float>(hkl.X());
+        millerindex[1] = static_cast<float>(hkl.Y());
+        millerindex[2] = static_cast<float>(hkl.Z());
+        inserter.insertMDEvent(
+            static_cast<float>(signal), static_cast<float>(error * error),
+            static_cast<uint16_t>(runnumber), detid, millerindex.data());
```

Comment:
The third parameter of `insertMDEvent` is not run number but run index (index into the vector of ExperimentInfo). Since it looks like you have only one `ExperimentInfo` then this value should be 0 not 1. This has been used incorrectly before and I tried to fix some of it in #20215.

Reply:
@rosswhitfield if I set it to 0, the unit test `test_DataWS()` fails
```
LoadDNSSCDTest.h:249: Error: Expected (events[i] == ref[i]) up to d (0.0000010000), found (0.0000000000 != 1.0000000008)
```

Reply:
Then the test is wrong. You will need to update `ref`.

Reply:
@rosswhitfield Thanks! I've commited the changes.